timeutil/schedule.go

```diff
@@ -233,3 +502,298 @@ func ParseLegacySchedule(scheduleSpec string) ([]*Schedule, error) {
+
+// parseWeekday will parse a string and extract weekday (eg. wed),
+// MonthWeekday (eg. 1st, 5th in the month).
+func parseWeekday(s string) (*Week, error) {
+	l := len(s)
+	if l != 3 && l != 4 {
+		return nil, fmt.Errorf("cannot parse %q: invalid format", s)
+	}
+
+	day := s
+	var pos uint
+	if l == 4 {
+		day = s[0:3]
+		if v, err := strconv.ParseUint(s[3:], 10, 32); err != nil {
+			return nil, fmt.Errorf("cannot parse %q: invalid week", s)
+		} else if v < 1 || v > 5 {
+			return nil, fmt.Errorf("cannot parse %q: incorrect week number", s)
+		} else {
+			pos = uint(v)
+		}
+	}
+
+	if !IsValidWeekday(day) {
+		return nil, fmt.Errorf("cannot parse %q: invalid weekday", s)
+	}
+
+	week := &Week{
+		Day: weekdayMap[day],
+		Pos: pos,
+	}
+	return week, nil
+}
+
+// parseCount will parse the string containing a count token and return the
+// count, the rest of the string with count information removed or an error
+func parseCount(s string) (count uint, rest string, err error) {
+	if strings.Contains(s, countToken) {
+		//     timespan = time ( "-" / "~" ) time [ "/" ( time / count ) ]
+		ws := strings.Split(s, countToken)
+		rest = ws[0]
+		countStr := ws[1]
+		c, err := strconv.ParseUint(countStr, 10, 32)
+		if err != nil || c == 0 {
+			return 0, "", fmt.Errorf("cannot parse %q: not a valid event interval",
+				s)
+		}
+		return uint(c), rest, nil
+	}
+	return 0, s, nil
+}
+
+// isValidWeekdaySpec returns true if string is of the following format:
+//     wday =  ( "sun" / "mon" / "tue" / "wed" / "thu" / "fri" / "sat" ) [ DIGIT ]
+func isValidWeekdaySpec(s string) bool {
+	_, err := parseWeekday(s)
+	return err == nil
+}
+
+const (
+	spanToken           = "-"
+	randomizedSpanToken = "~"
+	countToken          = "/"
+)
+
+var specialTokens = map[string]string{
+	// shorthand variant of whole day
+	"-": "0:00-24:00",
+	// and randomized whole day
+	"~": "0:00~24:00",
+}
+
+// Parse each event and return a slice of schedules.
+func parseEventSet(s string) (*Schedule, error) {
+	var events []string
+	// split event set into events
+	//     eventset = wdaylist / timelist / wdaylist "," timelist
+	// or wdaysets
+	//     wdaylist = wdayset *( "," wdayset )
+	// or timesets
+	//     timelist = timeset *( "," timeset )
+	//
+	// NOTE: the syntax is ambiguous in the sense the type of a 'set' is now
+	// explicitly indicated, thus the parsing will first try to handle it as
+	// a wdayset, then as timeset
+
+	if els := strings.Split(s, ","); len(els) > 1 {
+		events = els
+	} else {
+		events = []string{s}
+	}
+
+	var schedule Schedule
+	// indicates that any further events must be time events
+	var expectTime bool
+
+	for _, event := range events {
+		// process events one by one
+
+		randomized := false
+		var count uint
+		var start string
+		var end string
+
+		if c, rest, err := parseCount(event); err != nil {
+			return nil, err
+		} else if c > 0 {
+			// count was specified
+			count = c
+			// count token is only allowed in timespans, meaning
```

Comment:
I'll stop reviewing the logic here as the suggestion above will significantly change this, hopefully simplifying it. We can tell we only expect times because we got the first time, and we can tell we got the first time because it contains a : and we parsed it. The logic here feels a bit too complex for what is going on. I'd expect more like split, parseWeekSpan, parseWeekSpan, parseTimeSpan, parseTimeSpan, over.